riot.py

```diff
@@ -0,0 +1,480 @@
+"""
+TODO
+- json report for verify_all_tests
+- timing metadata (how long it takes to setup and run a suite, case, etc)
+- format/naming convention for cases
+- parallelize virtualenv creation - there's no reason to not to
+- intermediate venv building eg. if cases share deps X, Y, Z, generate an env
+  for these and copy it
+- does specifying encoding="utf-8" affect the locale of the test cases?
+- reporting to datadog (stats, traces)
+- typing
+- clean up into classes w/ methods?
+- better log formatter / output
+- better handling of CmdFailure in run_suites
+- add command to run pytest (or any command) in a venv
+  eg: python riot.py --suite tracer --python 3.7 --command pytest -k test_tracer_fork
+"""
+import argparse
+import itertools
+import logging
+import os
+import re
+import shutil
+import subprocess
+import sys
+import typing as t
+
+
+logger = logging.getLogger(__name__)
+
+
+SHELL = "/bin/bash"
+ENCODING = "utf-8"
+
+
+class AttrDict(dict):
+    def __init__(self, *args, **kwargs):
+        super(AttrDict, self).__init__(*args, **kwargs)
+        self.__dict__ = self
+
+
+class CmdFailure(Exception):
+    def __init__(self, msg, completed_proc):
+        self.msg = msg
+        self.proc = completed_proc
+        self.code = completed_proc.returncode
+        super().__init__(self, msg)
+
+
+# It's easier to read `Suite`, `Case` rather than AttrDict or dict.
+CaseInstance = Case = Suite = AttrDict
```

Comment:
Having custom classes such as:
```
class Case(AttrDict): pass
```
Is going to be a little bit better when introspecting or debugging.

Reply:
yepyepyep definitely, a lot of the functions laying around are really "methods" as well. This is purely just a result of iterating a ton and not wanting to deal with class boilerplate 🙂